### DISCUSSION TITLE
Selecting columns based on its datatype

### DISCUSSION THREAD

**Discussion by nilaymk** [ClickHouse/ClickHouse#41226] [Q&A]
I have a large de-normalised table (wide with ~150 columns) and I am only interested in columns of `String` and `Int64` types where I want count `uniq` values of these columns.

- I can't do `SELECT COLUMNS(reg_ex) APPLY(uniq) FROM my_db.my_table` as the column names don't have info of the data's type and hence no regular expression will be suitable.

- I can't specify individual columns like:
  `SELECT col1, col2, ...  APPLY(uniq) FROM my_db.my_table` OR
  `SELECT * EXCEPT(col10, col11, col21) ...  APPLY(uniq) FROM my_db.my_table`
  because:
  - I have too many columns for my sanity :-)
  - Since it is an evolving product, new columns are added to the table frequently. So each time I have to update the query manually.

**my question**
Is there a way of `SELECT`ing columns based on its data type?

e.g. in form of `SELECT 'all my String and Int64 columns' APPLY(uniq) FROM ...` OR

```
WITH (
  SELECT groupArray(name)
  FROM system.columns
  WHERE database='my_db'
    AND table='my_table'
    AND type IN ('String', 'Int64')
) as desired_columns
SELECT **COLUMNS(desired_columns)** FROM my_db.my_table
```
i.e Is there something I can replace `**COLUMNS(desired_columns)**` so that `desired_columns` get expanded to column names to choose the columns dynamically?

**Comment by mo-avatar** [comment]
Maybe you can filter the columns which you want based on type from system.columns table  and then processs it some how to form the sql you want. Just a little more work in application will do, instead of changing the clickhouse kernel it self.

**Comment by mneedham** [comment]
I'm trying to solve a similar problem and I think it'd be cool if you could pass in a lambda to the `COLUMNS` function (like how `APPLY` takes one) where you could apply arbitrary filtering logic. 

```
from 'yellow tripdata Jan 2023.parquet' 
select columns(col -> toTypeName(col) == 'Float64') 
LIMIT 1 
SETTINGS schema_inference_make_columns_nullable = 0;
```

Ideally we don't have to refer to the systems table, because that would only work for tables that have already been created, whereas in my case above I want to filter a Parquet file.

**Reply by UnamedRus** [reply]
Related https://github.com/ClickHouse/ClickHouse/issues/33685
